### ISSUE TITLE
List.delete ?

### ISSUE THREAD

**Issue by mechvel** [agda/agda-stdlib#988]
Consider
``` 
delete : (p? : Decidable P) → List A → List A
                                             -- delete the first x such that P x
```
What standard library has for this?

**Comment by gallais** [comment]
I don't think so. You could use [`Any`'s `_─_`](https://agda.github.io/agda-stdlib/Data.List.Relation.Unary.Any.html#2197)
together with [`Any`'s `any`](https://agda.github.io/agda-stdlib/Data.List.Relation.Unary.Any.html#2832).

**Comment by mechvel** [comment]
``Any._-_`` is good.
But the above ``delete`` is not given  ``Any P xs``, so that ``delete p? xs`` sometimes returns ``xs``.
May be ``delete`` is worth adding, I do not know.

**Comment by MatthewDaggitt** [comment]
@mechvel as always feel free to open a PR.

**Comment by jamesmckinna** [comment]
Cf. #989 
`All.decide` was added in https://github.com/agda/agda-stdlib/pull/1595 so you could in v2.1 define
```agda
module _ (P? : Decidable P) where

  private
    helper : ∀ xs → All (∁ P) xs ⊎ Any P xs
    helper = All.decide (Sum.swap ∘ (Dec.toSum ∘ P?))

  delete : List _ → List _
  delete xs with helper xs
  ... | inj₁ _ = xs
  ... | inj₂ p = xs Any.─ p
```

But it's not clear (to me, at least) which, if any, of `helper` or `delete` belongs in the library, by contrast with `All.decide`?
Suggest closing this issue?

**Comment by mechvel** [comment]
Personally I do not know. Currently I use my home-made functions
```
delP :  {xs : List A} → Any P xs → List A
delP {_ ∷ xs} (here _)        =  xs
delP {x ∷ xs} (there anyP-xs) =  x ∷ (delP anyP-xs)

delBy : (P? : Decidable P) → List A → List A
                                -- Delete the first x in xs such that P x, or return  xs  if there is not such x.
delBy _  []        =  []
delBy P? (x ∷ xs)  with P? x
... | yes _ =  xs
... | no _  =  x ∷ (delBy P? xs)
```
If the standard library announces an adequate  replacement for  them, then I would use the standard functions.

**Comment by jamesmckinna** [comment]
> ... Currently I use my home-made functions
> 
> ```
> delP :  {xs : List A} → Any P xs → List A
> delP {_ ∷ xs} (here _)        =  xs
> delP {x ∷ xs} (there anyP-xs) =  x ∷ (delP anyP-xs)

You should find (and be able to prove ;-)) that this is extensionally equivalent to `Any._─_` as previously suggested, so it's not obvious to me that you gain anything from this (slightly more direct) definition, which is the *fusion* of the two functions `removeAt` and `index` defining the library version...

**Comment by mechvel** [comment]
Sorry, `index` can mean in stdlib something different from a number.
It may be a proof  ``(i : x ∈ xs)``  for membership to a list.

**Comment by MatthewDaggitt** [comment]
I would be happy to have `delBy` added to `Data.List.Base` under the name `deleteFirst`.

As @jamesmckinna says, `delP` already exists.

@jamesmckinna I'm not so keen on your version of `delete` as it appears to traverse the list twice?

**Comment by mechvel** [comment]
Let the team decide. 
Anyway, double reversing (if any) will be bad here. Because `deleteFirst` will be run by interpreter or in executable,
and it is easy to avoid double reversing.

**Comment by jamesmckinna** [comment]
Agree with @MatthewDaggitt that 
* proposed `delete` should better be named `deleteFirst`
* the specimen implementation I offered is clearly excessive (written while I was thinking about #989 and trying to exploit `helper` there), while @mechvel 's `delBy` is a 'clean' definition

**Comment by mechvel** [comment]
I recall now of   Data.List.Relation.Unary.Any : 
```
_─_ : {P : Pred A p} → ∀ xs → Any P xs → List A
xs ─ x∈xs = removeAt xs (index x∈xs)
```
I see now that `index` indeed finds a number in `(Fin ...)` . 
So,  `_─_ ` first finds the number index for `x` , and then applies ``removeAt xs (index x∈xs)`` .
Do I understand correct that  for ``xs  = [0 .. n] = upTo n``  and  ``n∈xs :  n ∈ xs``
the call  ``xs   ─  n∈xs``   takes `2*n` steps to evaluate?
If so, then  it is better to change the implementation of  `_─_ `  to the sample of 
```
delP {_ ∷ xs} (here _)        =  xs
delP {x ∷ xs} (there anyP-xs) =  x ∷ (delP anyP-xs)
```
Because the latter takes n steps in the above example, and the function `_─_ `  is not only for proofs, but also for interpreted evaluation and for running in executable.
?
